libsolidity/analysis/FunctionCallGraph.h

```diff
@@ -0,0 +1,115 @@
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/ast/ASTVisitor.h>
+
+#include <vector>
+#include <variant>
+
+namespace solidity::frontend
+{
+
+/* Creates a Function call graph for a contract.
+ * Includes the following special nodes:
+ *  - CreationRoot: All calls made at contract creation originate from this node
+ *  - CreationDispatch: Represents the creation dispatch function
+ *  - ExternalDispatch: Represents the runtime dispatch for all external functions
+ *  - InternalDispatch: Represents the runtime dispatch for internal function pointers and complex expressions
+ *
+ *  Nodes are a variant of either the enum SpecialNode or an ASTNode pointer.
+ *  ASTNodes are usually inherited from CallableDeclarations
+ *  (FunctionDefinition, ModifierDefinition, EventDefinition) but for functions
+ *  without declaration it is directly the FunctionCall AST node.
+ *
+ *  Functions that are not called right away as well as functions without
+ *  declarations have an edge to the dispatch node. */
+class FunctionCallGraphBuilder: private ASTConstVisitor
+{
+public:
+	enum class SpecialNode { Unset, CreationRoot, CreationDispatch, InternalDispatch, ExternalDispatch };
+
+	typedef std::variant<ASTNode const*, SpecialNode> Node;
+
+	struct CompareByID
+	{
+		using is_transparent = void;
+
+		bool operator()(Node const& _lhs, Node const& _rhs) const
+		{
+			if (_lhs.index() != _rhs.index())
+				return _lhs.index() < _rhs.index();
+
+			if (std::holds_alternative<SpecialNode>(_lhs))
+				return std::get<SpecialNode>(_lhs) < std::get<SpecialNode>(_rhs);
+			return std::get<ASTNode const*>(_lhs)->id() < std::get<ASTNode const*>(_rhs)->id();
+		}
+		bool operator()(Node const& _lhs, int64_t _rhs) const
+		{
+			solAssert(!std::holds_alternative<SpecialNode>(_lhs), "");
+
+			return std::get<ASTNode const*>(_lhs)->id() < _rhs;
+		}
+		bool operator()(int64_t _lhs, Node const& _rhs) const
+		{
+			solAssert(!std::holds_alternative<SpecialNode>(_rhs), "");
+
+			return _lhs < std::get<ASTNode const*>(_rhs)->id();
+		}
+	};
+
+	struct ContractCallGraph
+	{
+		ContractCallGraph(ContractDefinition const& _contract) :contract(_contract) {}
+
+		/// Contract for which this is the graph
+		ContractDefinition const& contract;
+
+		std::map<Node, std::set<ASTNode const*, ASTNode::CompareByID>, CompareByID> edges;
```

Comment:
Why is this not ``std::map<Node, std::set<Node>>`` (plus comparators)? Can it also be `std::set<Node, Node>`` (plus comparators)?

Reply:
> Why is this not std::map<Node, std::set<Node>>

Because the leaf nodes never can be a `SpecialNode` (because regular function call is possible to them) so I figured it's best to avoid all the extra `std::variant` code when I don't need it... simplifying the code.

> Can it also be `std::set<Node, Node>` (plus comparators)?

You mean `std::set<std::pair<Node, Node>>`? I am not sure why we'd want this, as we'd lose our fast lookup using the key

Reply:
How do you model calls through the internal dispatch if you do not allow special nodes at the tip of an edge?

Reply:
As normal calls I guess? A straight edge without the internal dispatch.

Reply:
But that's the point - you don't know where the call goes to, that's why we need the internal dispatch

Reply:
Looks like it has already been changed to `Node` in the code so I'm marking this as resolved.